File: cpp/game_of_life.cpp
#include <iostream>
#include <array>
#include <vector>
#include <cassert>
#include "simple_time.h"

#define X_LEN 100
#define Y_LEN 40
#define RUN_TIME 1000
#define UPDATE_EVERY 500

using namespace std;

template<typename T>
std::ostream& operator<<(std::ostream& o,std::vector<T> v){
	o<<"(";
	for(unsigned int i=0; i<v.size(); i++){
		o<<v[i];
		if(i<v.size()-1) o<<",";
	}
	return o<<")";
}

struct Point{
	unsigned int x,y;
	Point();
	Point(unsigned int,unsigned int);
};

ostream& operator<<(ostream& o,Point a){
	return o<<"("<<a.x<<","<<a.y<<")";
}

vector<Point> get_neighbors(Point a){
	vector<Point> v;
	if(a.x>0){
		v.push_back({a.x-1,a.y});//left
		if(a.y>0)v.push_back({a.x-1,a.y-1});//up left
		if(a.y<Y_LEN-1)v.push_back({a.x-1,a.y+1});//down left
	}
	if(a.x<X_LEN-1){
		v.push_back({a.x+1,a.y});//right
		if(a.y>0)v.push_back({a.x+1,a.y-1});//up right
		if(a.y<Y_LEN-1)v.push_back({a.x+1,a.y+1});//down right
	}
	if(a.y>0) v.push_back({a.x,a.y-1});//up
	if(a.y<Y_LEN-1) v.push_back({a.x,a.y+1});//down	
	return v;
}

Point::Point():x(0),y(0){}
Point::Point(unsigned int a,unsigned int b):x(a),y(b){}

struct Grid{
	private:
	array<array<bool,X_LEN>,Y_LEN> grid;
	
	public:
	void init(vector<Point>);
	void update();
	array<array<bool,X_LEN>,Y_LEN> get();
	Grid();
};

Grid::Grid():grid({{}}){
	for(unsigned int y=0; y<Y_LEN; y++){
		for(unsigned int x=0; x<X_LEN; x++){
			grid[y][x]=false;
		}
	}
}

array<array<bool,X_LEN>,Y_LEN> Grid::get(){
	return grid;
}

/*
RULES:
Any live cell with fewer than two live neighbours dies
Any live cell with two or three live neighbours lives on to the next generation.
Any live cell with more than three live neighbours dies
Any dead cell with exactly three live neighbours becomes a live cell
*/

bool update_val(const Point p,const array<array<bool,X_LEN>,Y_LEN> grid){
	bool alive=grid[p.y][p.x];
	vector<Point> v=get_neighbors(p);
	unsigned int neighbors=0;
	for(unsigned int i=0;i<v.size();i++){
		Point t=v[i];
		if(grid[t.y][t.x]) neighbors++;
	}
	if(alive){
		switch(neighbors){
			case 0:
			case 1:
				return false;
			case 2:
			case 3:
				return true;
			default:
				return false;
		}
	}
	else return neighbors==3;
}

void Grid::update(){
	array<array<bool,X_LEN>,Y_LEN> new_grid=grid;
	for(unsigned int y=0; y<Y_LEN; y++){
		for(unsigned int x=0; x<X_LEN; x++){
			new_grid[y][x]=update_val(Point{x,y},grid);
		}
	}
	grid=new_grid;
}

void Grid::init(vector<Point> v){
	for(unsigned int i=0; i<v.size(); i++){
		Point p=v[i];
		assert(p.x<X_LEN && p.y<Y_LEN);
		grid[p.y][p.x]=true;//!grid[p.y][p.x];
	}
}

std::ostream& operator<<(std::ostream& o,Grid a){
	for(unsigned int y=0; y<Y_LEN; y++){
		for(unsigned int x=0; x<X_LEN; x++){
			o<<a.get()[y][x];
		}
		o<<"\n";
	}
	return o;
}

std::string to_string(Grid a){
	string s;
	for(unsigned int y=0; y<Y_LEN; y++){
		for(unsigned int x=0; x<X_LEN; x++){
			bool b = a.get()[y][x];
			s += b ? "X" : " ";
		}
		s+="\n";
	}
	return s;
}

void run(){
	srand(time(NULL));
	Grid grid;
	vector<Point> random_points;
	for(unsigned int x=0; x<X_LEN; x++){
		for(unsigned int y=0; y<Y_LEN; y++){
			bool random=(rand() % 2);
			if(random)random_points.push_back(Point{x,y});
		}
	}
	//vector<Point> init_points{{4,5},{4,4},{4,3}};
	//grid.init(init_points);
	grid.init(random_points);
	//cout<<grid<<"\n";
	for(unsigned int i=0;i<RUN_TIME; ){
		if(get_time(Time_type::MILLISECONDS)%UPDATE_EVERY!=0) continue;
		i++;
		grid.update();
		string grid_str = to_string(grid);
		system("clear");
		cout<<grid_str<<"\n";
	}
}

int main(){
	run();
	return 0;
}

File: python/game_of_life.py
#!/usr/bin/env python

import sys, os, time
from random import randint

GRID_WIDTH = 190 #cells
GRID_HEIGHT = 65 #cells
DELAY = 0 #seconds 

class Point(object):
	def __init__(self,x,y):
		self.x = x
		self.y = y
			
	def __str__(self):
		return "(" + str(self.x) + "," + str(self.y) + ")"

class Cell(object):
	def __init__(self,point,alive):
		self.point = point
		self.alive = alive 
		
	def set(self,alive):
		self.alive = alive
	
	def __str__(self):
		return "X" if self.alive else " "
		
	@staticmethod
	def count_alive(cells):
		living = 0
		for cell in cells:
			if cell.alive:
				living += 1
		return living		

class Grid(object):
	def __init__(self):
		self.cells = [[Cell(Point(x,y),bool(randint(0, 1))) for x in range(0,GRID_WIDTH)] for y in range(0,GRID_HEIGHT)]
	
	def __str__(self):
		grid = ""
		for row in self.cells:
			for cell in row:
				grid += str(cell)
			grid += "\n"
		return grid
		
	def get_neighbors(self,point):
		neighbors = []
		if point.x > 0:
			neighbors.append(self.cells[point.y][point.x - 1])
			if point.y > 0:
				neighbors.append(self.cells[point.y - 1][point.x - 1])
			if point.y < GRID_HEIGHT - 1: 
				neighbors.append(self.cells[point.y + 1][point.x - 1])
		if point.x < GRID_WIDTH - 1:
			neighbors.append(self.cells[point.y][point.x + 1])
			if point.y > 0: 
				neighbors.append(self.cells[point.y - 1][point.x + 1])
			if point.y < GRID_HEIGHT - 1: 
				neighbors.append(self.cells[point.y + 1][point.x + 1])
		if point.y > 0:
			neighbors.append(self.cells[point.y - 1][point.x])
		if point.y < GRID_HEIGHT - 1:
			neighbors.append(self.cells[point.y + 1][point.x])
		return neighbors		

	def update(self):
		new_cells = [[Cell(Point(x,y),False) for x in range(0,GRID_WIDTH)] for y in range(0,GRID_HEIGHT)]
		for row in self.cells:
			for cell in row:
				living_neighbors = Cell.count_alive(self.get_neighbors(cell.point))
				if cell.alive:
					new_cells[cell.point.y][cell.point.x].set(living_neighbors == 2 or living_neighbors == 3)
				else:
					new_cells[cell.point.y][cell.point.x].set(living_neighbors == 3)
		self.cells = new_cells
					
#main
grid = Grid();
while True:
	print(grid)
	grid.update()
	time.sleep(DELAY)
	os.system('clear')
